src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java/CheckstyleJavaParser.g4

```diff
@@ -0,0 +1,858 @@
+      ELLIPSIS id
+    ;
+
+recordBody
+    : LCURLY recordBodyDeclaration* RCURLY
+    ;
+
+recordBodyDeclaration
+    : compactConstructorDeclaration
+    | classBodyDeclaration
+    ;
+
+compactConstructorDeclaration
+    : mods+=modifier* id constructorBlock
+    ;
+
+classExtends
+    : EXTENDS_CLAUSE type=typeType[false]
+    ;
+
+implementsClause
+    : LITERAL_IMPLEMENTS typeList
+    ;
+
+typeParameters
+    : LT typeParameter (COMMA typeParameter)* GT
+    ;
+
+typeParameter
+    : annotations[false] id typeUpperBounds?
+    ;
+
+typeUpperBounds
+    : EXTENDS_CLAUSE annotations[false] typeBound
+    ;
+
+typeBound
+    : typeBoundType (BAND typeBoundType)*
+    ;
+
+typeBoundType
+    : annotations[false] classOrInterfaceOrPrimitiveType arrayDeclarator*
+    ;
+
+enumDeclaration[List<ModifierContext> mods]
+    : ENUM id implementsClause?
+      enumBody
+    ;
+
+enumBody
+    : LCURLY enumConstants? COMMA? enumBodyDeclarations? RCURLY
+    ;
+
+enumConstants
+    : enumConstant (COMMA enumConstant)*
+    ;
+
+enumConstant
+    : annotations[true] id arguments? classBody?
+    ;
+
+enumBodyDeclarations
+    : SEMI classBodyDeclaration*
+    ;
+
+interfaceDeclaration[List<ModifierContext> mods]
+    : LITERAL_INTERFACE id typeParameters?
+      interfaceExtends?
+      permittedSubclassesAndInterfaces?
+      interfaceBody
+    ;
+
+interfaceExtends
+    : EXTENDS_CLAUSE typeList
+    ;
+
+classBody
+    : LCURLY classBodyDeclaration* RCURLY
+    ;
+
+interfaceBody
+    : LCURLY interfaceBodyDeclaration* RCURLY
+    ;
+
+classBodyDeclaration
+    : SEMI                                                                 #emptyClass
+    | LITERAL_STATIC? block                                                #classBlock
+    | mods+=modifier*
+      memberDeclaration[{((ClassDefContext) _localctx).mods}]              #classDef
+    ;
+
+memberDeclaration[List<ModifierContext> mods]
+    : recordDeclaration[mods]
+    | methodDeclaration[mods]
+    | fieldDeclaration[mods]
+    | constructorDeclaration[mods]
+    | interfaceDeclaration[mods]
+    | annotationTypeDeclaration[mods]
+    | classDeclaration[mods]
+    | enumDeclaration[mods]
+    ;
+
+methodDeclaration[List<ModifierContext> mods]
+    : typeParams=typeParameters? type=typeType[true] id
+      formalParameters (cStyleArrDec+=arrayDeclarator*)
+      throwsList?
+      methodBody
+    ;
+
+methodBody
+    : block
+    | SEMI
+    ;
+
+throwsList
+    : LITERAL_THROWS qualifiedNameList
+    ;
+
+constructorDeclaration[List<ModifierContext> mods]
+    : typeParameters? id formalParameters
+      throwsList? constructorBody=constructorBlock
+    ;
+
+fieldDeclaration[List<ModifierContext> mods]
+    : type=typeType[true]
+      variableDeclarators[$mods, $ctx.type]
+      SEMI
+    ;
+
+interfaceBodyDeclaration
+    : mods+=modifier* interfaceMemberDeclaration[$ctx.mods]
+    | SEMI
+    ;
+
+interfaceMemberDeclaration[List<ModifierContext> mods]
```

Comment:
The modifiers RuleContext is handled similarly to ANTLR4 grammar; namely, we pass the RuleContext to the appropriate rule to build the respective AST.